Library/Homebrew/test/utils_spec.rb

```diff
@@ -0,0 +1,293 @@
+      end
+    end
+  end
+
+  describe "#interactive_shell" do
+    let(:shell) { dir/"myshell" }
+
+    it "starts an interactive shell session" do
+      IO.write shell, <<-EOS.undent
+        #!/bin/sh
+        echo called > "#{dir}/called"
+      EOS
+
+      FileUtils.chmod 0755, shell
+
+      ENV["SHELL"] = shell
+
+      expect { interactive_shell }.not_to raise_error
+      expect(dir/"called").to exist
+    end
+  end
+
+  describe "#with_custom_locale" do
+    it "temporarily overrides the system locale" do
+      ENV["LC_ALL"] = "en_US.UTF-8"
+
+      with_custom_locale("C") do
+        expect(ENV["LC_ALL"]).to eq("C")
+      end
+
+      expect(ENV["LC_ALL"]).to eq("en_US.UTF-8")
+    end
+  end
+
+  describe "#run_as_not_developer" do
+    it "temporarily unsets HOMEBREW_DEVELOPER" do
+      ENV["HOMEBREW_DEVELOPER"] = "foo"
+
+      run_as_not_developer do
+        expect(ENV["HOMEBREW_DEVELOPER"]).to be nil
+      end
+
+      expect(ENV["HOMEBREW_DEVELOPER"]).to eq("foo")
+    end
+  end
+
+  describe "#which" do
+    let(:cmd) { dir/"foo" }
+
+    before(:each) do
+      FileUtils.touch cmd
+    end
+
+    it "returns the first executable that is found" do
+      cmd.chmod 0744
+      expect(which(File.basename(cmd), File.dirname(cmd))).to eq(cmd)
+    end
+
+    it "skips non-executables" do
+      expect(which(File.basename(cmd), File.dirname(cmd))).to be nil
+    end
+
+    it "skips malformed path and doesn't fail" do
+      # 'which' should not fail if a path is malformed
+      # see https://github.com/Homebrew/legacy-homebrew/issues/32789 for an example
+      cmd.chmod 0744
+
+      # ~~ will fail because ~foo resolves to foo's home and there is no '~' user
+      path = ["~~", File.dirname(cmd)].join(File::PATH_SEPARATOR)
+      expect(which(File.basename(cmd), path)).to eq(cmd)
+    end
+  end
+
+  describe "#which_all" do
+    let(:cmd1) { dir/"foo" }
+    let(:cmd2) { dir/"bar/foo" }
+    let(:cmd3) { dir/"bar/baz/foo" }
+
+    before(:each) do
+      (dir/"bar/baz").mkpath
+
+      FileUtils.touch cmd2
+
+      [cmd1, cmd3].each do |cmd|
+        FileUtils.touch cmd
+        cmd.chmod 0744
+      end
+    end
+
+    it "returns an array of all executables that are found" do
+      path = [
+        "#{dir}/bar/baz",
+        "#{dir}/baz:#{dir}",
+        "~baduserpath",
+      ].join(File::PATH_SEPARATOR)
+      expect(which_all("foo", path)).to eq([cmd3, cmd1])
+    end
+  end
+
+  specify "#which_editor" do
+    ENV["HOMEBREW_EDITOR"] = "vemate"
+    expect(which_editor).to eq("vemate")
+  end
+
+  specify "#gzip" do
+    Dir.mktmpdir do |path|
+      path = Pathname.new(path)
+      somefile = path/"somefile"
+      FileUtils.touch somefile
+      expect(gzip(somefile)[0].to_s).to eq("#{somefile}.gz")
+      expect(Pathname.new("#{somefile}.gz")).to exist
+    end
+  end
+
+  specify "#capture_stderr" do
+    expect(
+      capture_stderr do
+        $stderr.print "test"
+      end,
```

Comment:
A `do…end` inside parentheses looks pretty strange to me. Could this be done with braces in one or two lines?